components/nodemanager-service/tests/project_update_test.go

```diff
@@ -0,0 +1,2231 @@
+									Values:    []string{"env1"},
+								},
+							},
+						},
+					},
+				},
+			},
+			originalProjectIDs: []string{"project9"},
+			expectedProjectIDs: []string{"project3"},
+		},
+		{
+			description: "Environment - not updating project. wrong rule type",
+			projectsData: []*nodes.ProjectsData{
+				{Key: "environment", Values: []string{"env1"}},
+			},
+			projectRules: map[string]*iam_v2.ProjectRules{
+				"project3": {
+					Rules: []*iam_v2.ProjectRule{
+						{
+							Type: iam_v2.ProjectRuleTypes_EVENT,
+							Conditions: []*iam_v2.Condition{
+								{
+									Attribute: iam_v2.ProjectRuleConditionAttributes_ENVIRONMENT,
+									Values:    []string{"env1"},
+								},
+							},
+						},
+					},
+				},
+			},
+			originalProjectIDs: []string{"project9"},
+			expectedProjectIDs: []string{},
+		},
+		{
+			description: "Environment - Single rule two matching conditions",
+			projectsData: []*nodes.ProjectsData{
+				{Key: "environment", Values: []string{"env1"}},
+				{Key: "organization_name", Values: []string{"org1"}},
+			},
+			projectRules: map[string]*iam_v2.ProjectRules{
+				"project3": {
+					Rules: []*iam_v2.ProjectRule{
+						{
+							Type: iam_v2.ProjectRuleTypes_NODE,
+							Conditions: []*iam_v2.Condition{
+								{
+									Attribute: iam_v2.ProjectRuleConditionAttributes_ENVIRONMENT,
+									Values:    []string{"env1"},
+								},
+								{
+									Attribute: iam_v2.ProjectRuleConditionAttributes_CHEF_ORGANIZATION,
+									Values:    []string{"org1"},
+								},
+							},
+						},
+					},
+				},
+			},
+			originalProjectIDs: []string{"project9"},
+			expectedProjectIDs: []string{"project3"},
+		},
+		{
+			description: "Environment - Single rule, one non-matching condition",
+			projectsData: []*nodes.ProjectsData{
+				{Key: "environment", Values: []string{"env1"}},
+			},
+			projectRules: map[string]*iam_v2.ProjectRules{
+				"project3": {
+					Rules: []*iam_v2.ProjectRule{
+						{
+							Type: iam_v2.ProjectRuleTypes_NODE,
+							Conditions: []*iam_v2.Condition{
+								{
+									Attribute: iam_v2.ProjectRuleConditionAttributes_ENVIRONMENT,
+									Values:    []string{"env1"},
+								},
+								{
+									Attribute: iam_v2.ProjectRuleConditionAttributes_CHEF_ORGANIZATION,
+									Values:    []string{"org1"},
+								},
+							},
+						},
+					},
+				},
+			},
+			originalProjectIDs: []string{"project9"},
+			expectedProjectIDs: []string{},
+		},
+		{
+			description: "Environment - a rule's condition has two values for a field",
+			projectsData: []*nodes.ProjectsData{
+				{Key: "environment", Values: []string{"env1"}},
+			},
+			projectRules: map[string]*iam_v2.ProjectRules{
+				"project3": {
+					Rules: []*iam_v2.ProjectRule{
+						{
+							Type: iam_v2.ProjectRuleTypes_NODE,
+							Conditions: []*iam_v2.Condition{
+								{
+									Attribute: iam_v2.ProjectRuleConditionAttributes_ENVIRONMENT,
+									Values:    []string{"env1", "env2"},
```

Comment:
Good! But suggest use env3, env1, env2, i.e., put env1 *not* as first or last. Otherwise, it might accidentally be only looking at the first (or the last) slot, but no one is going to write code to just look at a middle slot. 😁